r/R/dplyr-funcs-datetime.R

```diff
@@ -189,6 +189,65 @@ register_bindings_datetime <- function() {
   })
 }
 
+register_bindings_duration <- function() {
+  register_binding("difftime", function(time1,
+                                        time2,
+                                        tz,
+                                        units = c("auto", "secs", "mins",
+                                                  "hours", "days", "weeks")) {
+    units <- match.arg(units)
+    if (units != "secs") {
+      abort("`difftime()` with units other than seconds not supported in Arrow")
+    }
+
+    if (!missing(tz)) {
+      warn("`tz` is an optional argument to `difftime()` in R and will not be used in Arrow")
+    }
+
+    time1 <- build_expr("cast", time1, options = cast_options(to_type = timestamp()))
+    time2 <- build_expr("cast", time2, options = cast_options(to_type = timestamp()))
+
+    build_expr("cast", time1 - time2, options = cast_options(to_type = duration("s")))
+  })
+
+  register_binding("as.difftime", function(x,
+                                           format = "%X",
+                                           units = "auto") {
+    # windows doesn't seem to like "%X"
+    if (format == "%X" & tolower(Sys.info()[["sysname"]]) == "windows") {
+      format <- "%H:%M:%S"
+    }
+
+    if (units != "secs") {
+      abort("`as.difftime()` with units other than seconds not supported in Arrow")
+    }
+
+    if (call_binding("is.character", x)) {
+      x <- build_expr("strptime", x, options = list(format = format, unit = 0L))
+      y <- build_expr("strptime", "0:0:0", options = list(format = "%H:%M:%S", unit = 0L))
+      diff_x_y <- call_binding("difftime", x, y, units = "secs")
+      return(diff_x_y)
+    }
```

Comment:
Have you looked into doing strptime on `x` and then converting to `time32()`? I think that + casting to duration will do what you want it to, without needing to construct a separate `y` here.

Reply:
Neither `time32()`, nor `time64()` can be cast to `duration`.

Reply:
Do we have a jira for allowing them to? They probably should be able to be cast to `duration`s

Reply:
Jira - [ARROW-15996](https://issues.apache.org/jira/browse/ARROW-15996) + `reprex` below

``` r
library(arrow, warn.conflicts = FALSE)

a <- Array$create(32000L)
a
#> Array
#> <int32>
#> [
#>   32000
#> ]
a$cast(time32())
#> Array
#> <time32[ms]>
#> [
#>   00:00:32.000
#> ]
a$cast(int64())$cast(time64())
#> Array
#> <time64[ns]>
#> [
#>   00:00:00.000032000
#> ]

a$cast(time32())$cast(duration())
#> Error: NotImplemented: Unsupported cast from time32[ms] to duration using function cast_duration
a$cast(int64())$cast(time64())$cast(duration())
#> Error: NotImplemented: Unsupported cast from time64[ns] to duration using function cast_duration
```

<sup>Created on 2022-03-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
I've done a bit more thinking and digging around this. It turns out {arrow} maps `time32()` to R's `difftime`. So, in theory, we don't need `y` and we could simply cast `x` to `time32()`. In practice it's not as straightforward since `x` also inherits the class of the object it transforms into `difftime`. In the example below, that class is `hms`:

``` r
library(arrow, warn.conflicts = FALSE)
library(dplyr, warn.conflicts = FALSE)
library(lubridate, warn.conflicts = FALSE)

test_df <- tibble(
  hms_string = c("0:7:45", "12:34:56"),
  hm_string = c("7:45", "12:34"),
  int = c(30L, 75L),
  integerish_dbl = c(31, 76),
  dbl = c(31.2, 76.4)
)

test_df %>%
  mutate(hms_difftime = as.difftime(hms_string, units = "secs"))
#> # A tibble: 2 × 6
#>   hms_string hm_string   int integerish_dbl   dbl hms_difftime
#>   <chr>      <chr>     <int>          <dbl> <dbl> <drtn>      
#> 1 0:7:45     7:45         30             31  31.2   465 secs  
#> 2 12:34:56   12:34        75             76  76.4 45296 secs

test_df %>%
  arrow_table() %>%
  mutate(hms_difftime = as.difftime(hms_string, units = "secs")) %>%
  collect()
#> # A tibble: 2 × 6
#>   hms_string hm_string   int integerish_dbl   dbl hms_difftime
#>   <chr>      <chr>     <int>          <dbl> <dbl> <drtn>      
#> 1 0:7:45     7:45         30             31  31.2   465 secs  
#> 2 12:34:56   12:34        75             76  76.4 45296 secs

compare_dplyr_binding(
    .input %>%
      mutate(hms_difftime = as.difftime(hms_string, units = "secs")) %>%
      collect(),
    test_df
  )
#> Error: `object` (`actual`) not equal to `expected` (`expected`).
#>
#> `class(actual$hms_difftime)`: "hms" "difftime"
#> `class(expected$hms_difftime)`: "difftime"
```

<sup>Created on 2022-03-23 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

We could simply ignore attributes but that doesn't feel right.

Reply:
I think we have at least 3 options for this issue:

1. `x` only - use return `time32()` which currently is mapped to `hms::difftime`. For version `v7.0.0` this mapping is documented, for `v 8.0.0` I think the documentation doesn't mention it (as we now have `duration <-> difftime`. I personally am not a huge fan of this since we would need to rely on un-documented behaviour and we have a slight mismatch in class.

2. `x` only - go through quite a long cycle of casting to get to duration with the correct size. We need to cast twice to `int64()` and need to account for the shift in measurement unit by dividing by 10^9. This seems cumbersome and a bit difficult to read. 
```r
x <- build_expr("strptime", x, options = list(format = format, unit = 0L))
x <- build_expr("/", x$cast(time64())$cast(int64()), Scalar$create(10^9, int64()))
x <- x$cast(int64())$cast(duration("s"))
```

3. `x` and `y`. Subtract `"0:0:0"` from x (my initial approach). I still think this is the most elegant solution.

Reply:
Have you seen https://issues.apache.org/jira/browse/ARROW-15996?focusedCommentId=17511259&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17511259 ? 

Would the following work?
```
# strptime, convert to time64
x <- build_expr("strptime", x, options = list(format = format, unit = 0L))$cast(time64("us"))
# Complex casting only due to cast type restrictions: time64 -> int64 -> duration(us) -> duration(s)
x <- x$cast(int64())$cast(duration("us"))$cast(duration("s"))
```

Yes, it does involve a series of casts (though most of those should be pretty cheap since it's mostly integers being moved around. 

The nice thing about this is that it's conceptually isolated: going from a "%H:%M:%S" to a time64 to a duration. Where as (3) requires a few assumptions about how `strptime()` fills in date details (if you check this it's actually totally different in Arrow and R!) and then discards them. And agree that (1) is not ideal: `difftime` in R is definitely more like Arrow's `duration` than `time32|64()` objects. 

And there are a number of ways this sequence of casts could be simplified (all the examples below use the array: `arr <- Array$create("01:00:00")`) either in ARROW-15996 or elsewhere: 
* being able to go from `time64` to `duration` directly: 
    `call_function("strptime", arr, options = list(format = "%H:%M:%S", unit = 0L))$cast(time64("us"))$cast(duration("s"))`
* being able to go from `int32` to duration, then we could do something like: 
    `call_function("strptime", arr, options = list(format = "%H:%M:%S", unit = 0L))$cast(time32("s"))$cast(int32())$cast(duration("s"))`
* Probably a few more...

Reply:
I've updated the casting chain with your comment. I think this is good go.